feat(api): implement liquid probe protocol engine command

pmoegenburg opened the pull request:
<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview

<!--
Use this section to describe your pull-request at a high level. If the PR addresses any open issues, please tag the issues here.
-->

This closes EXEC-436. Pulls well dimension data from Protocol Engine StateView, which is now used for process `max_z_distance`. Also bubbles up LiquidNotFoundError.

# Test Plan

<!--
Use this section to describe the steps that you took to test your Pull Request.
If you did not perform any testing provide justification why.

OT-3 Developers: You should default to testing on actual physical hardware.
Once again, if you did not perform testing against hardware, justify why.

Note: It can be helpful to write a test plan before doing development

Example Test Plan (HTTP API Change)

- Verified that new optional argument `dance-party` causes the robot to flash its lights, move the pipettes,
then home.
- Verified that when you omit the `dance-party` option the robot homes normally
- Added protocol that uses `dance-party` argument to G-Code Testing Suite
- Ran protocol that did not use `dance-party` argument and everything was successful
- Added unit tests to validate that changes to pydantic model are correct

-->

Added protocol engine command test.

# Changelog

<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

# Review requests

<!--
Describe any requests for your reviewers here.
-->

# Risk assessment

<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

Low. Largely builds off existing Aspirate protocol engine command.

pmoegenburg committed:
placeholder file

pmoegenburg committed:
initial commit

sfoster1 reviewed (changes requested):
Aside from fixing what's already failing tests, please either draft this or take a pass to remove note to self comments and write some tests. They can be pretty simple for now since this command doesn't do much, but they should be present.

pmoegenburg committed:
build fix updates

ryanthecoder commented on api/src/opentrons/protocol_engine/commands/liquid_probe.py:
```diff
@@ -0,0 +1,87 @@
+class LiquidProbeParams(BaseModel):
+    """Payload required to liquid probe."""
+
+    mount: MountType = Field(..., description="Instrument mount to liquid probe with.")
```
We are going to want the well we are using so we can get the well's max depth

ryanthecoder commented on api/src/opentrons/protocol_engine/commands/liquid_probe.py:
```diff
@@ -0,0 +1,87 @@
+            ot3_mount is not OT3Mount.GRIPPER
+        ), "Expected a Pipette mount but Gripper mount was provided."
+
+        z_pos = ot3_api.liquid_probe(mount=ot3_mount)  # anything else?
```
We are going to want to pass an Instance of LiquidProbeSettings.
we can get a copy of the defaults from ot3_api.config().liquid_sense
We are going to want to change the starting mount height and the max_z_distance based on the well we are probing

pmoegenburg committed:
building except tests

pmoegenburg committed:
test fixes

pmoegenburg committed:
cleaned up comments

pmoegenburg committed:
updated command process

pmoegenburg committed:
pass in max_z_dist

pmoegenburg committed:
Merge branch 'edge' into EXEC-436-lld-protocol-engine-command

caila-marashaj commented on api/src/opentrons/protocol_engine/commands/liquid_probe.py:
```diff
@@ -0,0 +1,116 @@
+        )
+
+        return SuccessData(
+            public=LiquidProbeResult(
```
I think it might just be worth a comment somewhere that the `position` here is the position at the well where the probe starts, and `z_pos` is after the probe

caila-marashaj reviewed (approved):
Nice!

pmoegenburg committed:
test fixes

pmoegenburg committed:
Merge branch 'edge' into EXEC-436-lld-protocol-engine-command

pmoegenburg committed:
all existing tests passing

pmoegenburg committed:
bubbled up LiquidNotFoundError

pmoegenburg committed:
updated command tests

pmoegenburg committed:
Merge branch 'edge' into EXEC-436-lld-protocol-engine-command

pmoegenburg committed:
method call fix

pmoegenburg committed:
another method call fix

ryanthecoder commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -2622,21 +2623,21 @@ async def liquid_probe(
+        pos = await self.gantry_position(checked_mount, refresh=True)
         probe_start_pos = pos._replace(z=probe_settings.starting_mount_height)
-        await self.move_to(mount, probe_start_pos)
-        total_z_travel = probe_settings.max_z_distance
```
If we're not going to use this value from probe_settings. Lets remove it from that data struct so we don't confuse anything or miss any references using it in hardware-testing

sfoster1 commented on api/src/opentrons/hardware_control/protocols/liquid_handler.py:
```diff
@@ -171,3 +171,15 @@ async def drop_tip(
+        """Search for and return liquid level height using this pipette
+        at the current location.
+
+        mount : Mount.LEFT or Mount.RIGHT
```
should have `max_z_dist` explained too

sfoster1 commented on api/src/opentrons/protocol_engine/commands/liquid_probe.py:
```diff
@@ -0,0 +1,152 @@
+class LiquidProbeResult(DestinationPositionResult):
+    """Result data from the execution of a liquid-probe command."""
+
+    z_position: float = Field(..., description="Z position of the found liquid.")
```
worth mentioning here that
- This is the _absolute_ z position, as in from the deck, not the relative height from the bottom of the well
- It is in mm

sfoster1 commented on api/src/opentrons/protocol_engine/execution/pipetting.py:
```diff
@@ -245,6 +271,15 @@ async def blow_out_in_place(
+        well_name: str,
+    ) -> float:
+        """Detect liquid level."""
+        return 0.0  # fix
```
let's either do the fix now or let's make it a better formatted TODO comment

sfoster1 reviewed (changes requested):
Couple last little changes but almost there!

pmoegenburg committed:
eliminated max_z_distance from LiquidProbeSettings struct

pmoegenburg committed:
PR review and lint fixes

pmoegenburg committed:
formatted

pmoegenburg committed:
Merge branch 'edge' into EXEC-436-lld-protocol-engine-command

github-actions[bot] committed:
fix(app-testing): snapshot failure capture (#15455)

This PR is an automated snapshot update request. Please review the
changes and merge if they are acceptable or find you bug and fix it.

Co-authored-by: pmoegenburg <pmoegenburg@users.noreply.github.com>
Co-authored-by: pmoegenburg <pbm53@cornell.edu>

sfoster1 reviewed (approved):
Very nice, love the changes, looks good to me!
